include/rmm/mr/device/detail/stream_ordered_memory_resource.hpp

```diff
@@ -288,17 +289,26 @@ class stream_ordered_memory_resource : public crtp<PoolResource>, public device_
   stream_event_pair get_event(cuda_stream_view stream)
   {
     if (stream.is_per_thread_default()) {
-      // Create a thread-local shared event wrapper. Shared pointers in the thread and in each MR
-      // instance ensures it is destroyed cleaned up only after all are finished with it.
-      thread_local auto event_tls = std::make_shared<event_wrapper>();
-      default_stream_events.insert(event_tls);
-      return stream_event_pair{stream.value(), event_tls->event};
+      // Create a thread-local shared event wrapper for each device. Shared pointers in the thread
+      // and in each MR instance ensure the wrappers are destroyed only after all are finished
+      // with them.
+      thread_local std::vector<std::shared_ptr<event_wrapper>> events_tls(
+        rmm::get_num_cuda_devices());
+      auto event = [&, device_id = this->device_id_]() {
+        if (events_tls[device_id.value()]) { return events_tls[device_id.value()]->event; }
+
+        auto event                    = std::make_shared<event_wrapper>();
```

Comment:
So this creates an event on the currently active device, so there's an implicit requirement that `get_event` is called with `cudaGetDevice` equal to `this->device_id_`. See discussion below on whether we want to ensure this internally, or document the requirements explicitly.

Reply:
I think we do document that a memory resource is only valid to use on the device where it was created. That means the device must be current before calling any memory resource methods. Next week I will seek out the documentation and make sure it's sufficient.